### PR TITLE
feat: _BASE_URL (for opentreehole), _BASE_AUTH_URL and _IMAGE_BASE_URL are now modifiable

### LINKED ISSUE #260: [Feature Request] 动态修改 opentreehole 的请求 URL
**你的功能需求和某个bug有关吗？**
否

**你想要什么样的功能？**
目前 opentreehole repository 的 _BASE_URL, _BASE_AUTH_URL, _IMAGE_BASE_URL 都是写死的 string，只能用与正式服，不方便使用测试服调试。希望可以提供一个动态修改这三个 URL 的页面。

### PR DESCRIPTION
**Pull request by ivanfei-1**
close #260.

### CODE REVIEW

File: lib/provider/settings_provider.dart [hunk 1]
```diff
@@ -79,6 +79,9 @@ class SettingsProvider with ChangeNotifier {
   static const String KEY_THEME_TYPE = "theme_type";
   static const String KEY_MARKDOWN_ENABLED = "markdown_rendering_enabled";
   static const String KEY_VISITED_TIMETABLE = "visited_timetable";
+  static const String KEY_BASE_URL = "https://www.fduhole.com/api";
+  static const String KEY_BASE_AUTH_URL = "https://auth.fduhole.com/api";
+  static const String KEY_IMAGE_BASE_URL = "https://image.fduhole.com";
```
Comment (w568w) [inline lib/provider/settings_provider.dart:84]: Do not use a site URL as the key. Follow the naming style of other keys (i.e. `lower_snake_case`).

File: lib/provider/settings_provider.dart [hunk 2]
```diff
@@ -129,6 +132,61 @@ class SettingsProvider with ChangeNotifier {
     }
   }
 
+  /// Set and get _BASE_URL, _BASE_AUTH_URL, _IMAGE_BASE_URL for debug
+  String get baseUrl {
+    final String? baseUrl = preferences!.getString(KEY_BASE_URL);
+    if (baseUrl != null) {
+      return baseUrl;
+    } else {
+      return "https://www.fduhole.com/api";
```
Comment (w568w) [inline lib/provider/settings_provider.dart:141]: Do not hard-code the base URLs everywhere. AFAICS, they are at least coded again and again, in:

1. `lib/page/settings/diagnostic_console.dart` (3x)
2. `lib/provider/settings_provider.dart` (6x)

You should either define the default value **only** in `constants.dart`, or **only** in `get XXX {}` methods here.

File: lib/provider/settings_provider.dart [hunk 3]
```diff
@@ -129,6 +132,61 @@ class SettingsProvider with ChangeNotifier {
...
+    }
+  }
+
+  set baseUrl(String? value) {
+    if (value != null) {
+      preferences!.setString(KEY_BASE_URL, value);
+    } else {
+      preferences!.setString(KEY_BASE_URL, "https://www.fduhole.com/api");
```
Comment (w568w) [inline lib/provider/settings_provider.dart:149]: Why set the url to its default value when someone calls `SettingsProvider.baseUrl = null`? 

IMMO this `set` should not accept a nullable parameter at all, i.e. 

```
set baseUrl(String value) {
// ...
}
```

**Review by w568w — CHANGES_REQUESTED** [review]

File: lib/common/constant.dart [hunk 4]
```diff
@@ -73,6 +73,12 @@ class Constant {
   static const String UIS_URL = "https://uis.fudan.edu.cn/authserver/login";
   static const String UIS_HOST = "uis.fudan.edu.cn";
 
+  /// The default urls of the OpenTreeHoleRepository
```
Comment (w568w) [inline lib/common/constant.dart:76]: It is recommended to follow the documentation syntax to highlight the names of classes and objects with `[]`, e.g.

```
The default URLs of [OpenTreeHoleRepository].
```

In IDEs (like Intellij IDEA), you can jump to `OpenTreeHoleRepository`'s definition by clicking on `[OpenTreeHoleRepository]` with `Command` pressed.

**Review by w568w — APPROVED** [review]
Almost great to me. Just a subtle change required.

File: lib/common/constant.dart (hunk 4 above)
Comment (ivanfei-1) [reply lib/common/constant.dart]: All done.

**Pull request merged**